packages/flutter_tools/lib/src/custom_devices/custom_device.dart

```diff
@@ -0,0 +1,725 @@
+  /// of `${localPath}` in the custom device configs `install` command.
+  ///
+  /// [appName] is the name of the app to be installed. Substituted for any occurrence
+  /// of `${appName}` in the custom device configs `install` command.
+  Future<bool> _tryInstall({
+    @required String localPath,
+    @required String appName,
+    Duration timeout,
+    Map<String, String> additionalReplacementValues = const <String, String>{}
+  }) async {
+    final List<String> interpolated = interpolateCommand(
+      _config.installCommand,
+      <String, String>{
+        'localPath': localPath,
+        'appName': appName
+      },
+      additionalReplacementValues: additionalReplacementValues
+    );
+
+    try {
+      await _processUtils.run(
+        interpolated,
+        throwOnError: true,
+        timeout: timeout
+      );
+
+      return true;
+    } on ProcessException catch (e) {
+      _logger.printError('Error executing install command for custom device $id: $e');
+      return false;
+    }
+  }
+
+  @override
+  void clearLogs() {}
+
+  @override
+  Future<void> dispose() async {
+    _sessions
+        ..forEach((_, CustomDeviceAppSession session) => session.dispose())
+        ..clear();
+  }
+
+  @override
+  Future<String> get emulatorId async => null;
+
+  @override
+  FutureOr<DeviceLogReader> getLogReader({
+    covariant ApplicationPackage app,
+    bool includePastLogs = false
+  }) {
+    if (app != null) {
+      return _getOrCreateAppSession(app).logReader;
+    } else {
+      return _globalLogReader;
+    }
+  }
+
+  @override
+  Future<bool> installApp(covariant ApplicationPackage app, {String userIdentifier}) async {
+    if (!await _tryUninstall(appName: app.name)) {
+      return false;
+    }
+
+    final bool result = await _tryInstall(
+      localPath: getAssetBuildDirectory(),
+      appName: app.name
+    );
+
+    return result;
+  }
+
+  @override
+  Future<bool> isAppInstalled(covariant ApplicationPackage app, {String userIdentifier}) {
+    return Future<bool>.value(false);
+  }
+
+  @override
+  Future<bool> isLatestBuildInstalled(covariant ApplicationPackage app) {
+    return Future<bool>.value(false);
+  }
+
+  @override
+  Future<bool> get isLocalEmulator async => false;
+
+  @override
+  bool isSupported() {
+    return true;
+  }
+
+  @override
+  bool isSupportedForProject(FlutterProject flutterProject) {
+    return true;
+  }
+
+  @override
+  FutureOr<bool> supportsRuntimeMode(BuildMode buildMode) {
+    return buildMode == BuildMode.debug;
+  }
+
+  @override
+  String get name => _config.label;
+
+  @override
+  DevicePortForwarder get portForwarder {
+    if (_config.usesPortForwarding) {
+      return CustomDevicePortForwarder(
+        deviceName: name,
+        forwardPortCommand: _config.forwardPortCommand,
+        forwardPortSuccessRegex: _config.forwardPortSuccessRegex,
+        processManager: _processManager,
+        logger: _logger,
+      );
+    } else {
```

Comment:
nit: remove the `else` since you have an early return above

Reply:
fixed it